median() doesn't work with an ARR64 view column

```
>>> D = dt.Frame(A=list(range(10)))[np.array(range(10)), :]
>>> D[:, dt.median(dt.f.A)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'i < nrows()' failed in src/core/column.cc, line 215

>>> D = dt.Frame(A=range(10))[np.array(range(10)), :]
>>> D[:, dt.median(dt.f.A)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Invalid data access for a virtual column
```
in production code this results in a crash.

_Originally reported by: @pseudotensor (https://github.com/h2oai/h2oai/issues/19239)_

--- a/docs/releases/v1.0.0.rst
+++ b/docs/releases/v1.0.0.rst
@@ -45,6 +45,10 @@
     -[fix] When a whole column is updated within a ``DT[i, j, by()]`` call,
       the stype/ltype of that column us now allowed to change. [#2685]
 
+    -[fix] Fix a crash that occurred when using `median()` on virtual
+      columns of type ArrayView64. [#2802]
+
+
     fread
     -----
 

--- a/src/core/column/view.cc
+++ b/src/core/column/view.cc
@@ -88,13 +88,13 @@ ArrayView_ColumnImpl<T>::ArrayView_ColumnImpl(
   : Virtual_ColumnImpl(nrows, col.stype()),
     arg(std::move(col))
 {
-  xassert((std::is_same<T, int32_t>::value? ri.isarr32() : ri.isarr64()));
-  xassert(ri.max() < arg.nrows());
   set_rowindex(ri);
 }
 
 template <typename T>
 void ArrayView_ColumnImpl<T>::set_rowindex(const RowIndex& ri) {
+  xassert((std::is_same<T, int32_t>::value? ri.isarr32() : ri.isarr64()));
+  xassert(ri.max() < arg.nrows());
   rowindex_container = ri;
   indices = get_indices<T>(ri);
 }

--- a/src/core/sort.cc
+++ b/src/core/sort.cc
@@ -1506,7 +1506,15 @@ void dt::ArrayView_ColumnImpl<T>::sort_grouped(
   (void) out.stats();
   SortContext sc(nrows(), rowindex_container, grps, /*make_groups=*/ false);
   sc.continue_sort(arg, /*desc=*/ false, /*make_groups=*/ false);
-  set_rowindex(sc.get_result_rowindex());
+  if (sizeof(T) == 4) {
+    set_rowindex(sc.get_result_rowindex());
+  } else {
+    // Since SortContext produces an ARR32 rowindex, and the current
+    // ArrayView_ColumnImpl is int64, we need to replace the column
+    // with the new which uses int32s.
+    out = Column(new dt::ArrayView_ColumnImpl<int32_t>(
+                  std::move(arg), sc.get_result_rowindex(), nrows_));
+  }
 }
 
 template class dt::ArrayView_ColumnImpl<int32_t>;

--- a/src/core/sort.cc
+++ b/src/core/sort.cc
@@ -585,6 +585,7 @@ class SortContext {
 
 
   void continue_sort(const Column& col, bool desc, bool make_groups) {
+    xassert(!col.is_virtual());
     column = col;
     nradixes = gg.size();
     descending = desc;
@@ -1504,6 +1505,7 @@ void dt::ArrayView_ColumnImpl<T>::sort_grouped(
     const Groupby& grps, Column& out)
 {
   (void) out.stats();
+  arg.materialize();
   SortContext sc(nrows(), rowindex_container, grps, /*make_groups=*/ false);
   sc.continue_sort(arg, /*desc=*/ false, /*make_groups=*/ false);
   if (sizeof(T) == 4) {

--- a/tests/test-reduce.py
+++ b/tests/test-reduce.py
@@ -434,6 +434,20 @@ def test_median_wrong_stype():
             "type str64" in str(e.value))
 
 
+def test_median_issue2802_1():
+    I = dt.Frame(list(range(10)), stype=dt.int64)
+    DT = dt.Frame(A=list(range(10)))[I, :]
+    RES = DT[:, median(f.A)]
+    assert_equals(RES, dt.Frame(A=[4.5]))
+
+
+def test_median_issue2802_2():
+    I = dt.Frame(list(range(13)), stype=dt.int64)
+    DT = dt.Frame(A=range(13))[I, :]
+    RES = DT[:, median(f.A)]
+    assert_equals(RES, dt.Frame(A=[6.0]))
+
+
 
 #-------------------------------------------------------------------------------
 # Cov

